### ISSUE TITLE
AttributeError when building from multiple sources

### ISSUE THREAD

**Issue by jlumpe** [conda/conda-build#2236]
I am attempting to build my recipe from multiple sources, but conda-build is raising an exception on the source step. My desired sources in `meta.yaml` look something like this:

```yaml
  - git_url: https://github.com/jlumpe/lyve-SET.git
    git_rev: v1.1.4-head
    patches:
      - lyve-set.patch

  - git_url: https://github.com/lskatz/Schedule--SGELK.git
    folder: build/Schedule

  - git_url: https://github.com/lskatz/cg-pipeline
    folder: libs/cg-pipeline

  - url: http://phast.wishartlab.com/phage_finder/DB/prophage_virus.db
    folder: build

  - path: templates
    folder: build/templates
```

Running `conda-build -s .` in the directory with my meta.yaml file I get the following error:

```
Traceback (most recent call last):
  File "/home/jared/anaconda3/bin/conda-build", line 6, in <module>
    sys.exit(conda_build.cli.main_build.main())
  File "/home/jared/anaconda3/lib/python3.5/site-packages/conda_build/cli/main_build.py", line 342, in main
    execute(sys.argv[1:])
  File "/home/jared/anaconda3/lib/python3.5/site-packages/conda_build/cli/main_build.py", line 329, in execute
    outputs = [action(recipe, config) for recipe in args.recipe]
  File "/home/jared/anaconda3/lib/python3.5/site-packages/conda_build/cli/main_build.py", line 329, in <listcomp>
    outputs = [action(recipe, config) for recipe in args.recipe]
  File "/home/jared/anaconda3/lib/python3.5/site-packages/conda_build/cli/main_build.py", line 272, in source_action
    source.provide(metadata)
  File "/home/jared/anaconda3/lib/python3.5/site-packages/conda_build/source.py", line 608, in provide
    path = normpath(abspath(join(metadata.path, metadata.get_value('source/path'))))
  File "/home/jared/anaconda3/lib/python3.5/site-packages/conda_build/metadata.py", line 891, in get_value
    value = self.get_section(section).get(key, default)
AttributeError: 'list' object has no attribute 'get'
```

It looks like it is trying to treat the `source` key as a dictionary instead of recognizing it is a list of dictionaries. The strange thing is various subsets of these sources will work or not work.

Here is a very basic version that reproduces the problem:

```yaml
source: 
  - git_url: https://github.com/jlumpe/lyve-SET.git
  - path: templates
```

Excluding only the local source works when using `conda-build -s`. However it fails immediately after getting the sources when doing the full build, even when I limit it to only the git sources:

```yaml
source:
  - git_url: https://github.com/jlumpe/lyve-SET.git
    git_rev: v1.1.4-head

  - git_url: https://github.com/lskatz/Schedule--SGELK.git
    folder: build/Schedule

  - git_url: https://github.com/lskatz/cg-pipeline
    folder: libs/cg-pipeline
```

```yaml
Traceback (most recent call last):
  File "/home/jared/anaconda3/bin/conda-build", line 6, in <module>
    sys.exit(conda_build.cli.main_build.main())
  File "/home/jared/anaconda3/lib/python3.5/site-packages/conda_build/cli/main_build.py", line 342, in main
    execute(sys.argv[1:])
  File "/home/jared/anaconda3/lib/python3.5/site-packages/conda_build/cli/main_build.py", line 333, in execute
    noverify=args.no_verify)
  File "/home/jared/anaconda3/lib/python3.5/site-packages/conda_build/api.py", line 184, in build
    need_source_download=need_source_download, config=config, variants=variants)
  File "/home/jared/anaconda3/lib/python3.5/site-packages/conda_build/build.py", line 1633, in build_tree
    notest=notest,
  File "/home/jared/anaconda3/lib/python3.5/site-packages/conda_build/build.py", line 999, in build
    m.parse_until_resolved(allow_no_other_outputs=True)
  File "/home/jared/anaconda3/lib/python3.5/site-packages/conda_build/metadata.py", line 824, in parse_until_resolved
    bypass_env_check=bypass_env_check)
  File "/home/jared/anaconda3/lib/python3.5/site-packages/conda_build/metadata.py", line 759, in parse_again
    bypass_env_check=bypass_env_check),
  File "/home/jared/anaconda3/lib/python3.5/site-packages/conda_build/metadata.py", line 1277, in _get_contents
    bypass_env_check=bypass_env_check))
  File "/home/jared/anaconda3/lib/python3.5/site-packages/conda_build/jinja_context.py", line 412, in context_processor
    ctx = get_environ(config=config, m=initial_metadata, for_env=False)
  File "/home/jared/anaconda3/lib/python3.5/site-packages/conda_build/environ.py", line 257, in get_dict
    d.update(meta_vars(m, config))
  File "/home/jared/anaconda3/lib/python3.5/site-packages/conda_build/environ.py", line 383, in meta_vars
    git_url = meta.get_value('source/git_url')
  File "/home/jared/anaconda3/lib/python3.5/site-packages/conda_build/metadata.py", line 891, in get_value
    value = self.get_section(section).get(key, default)
AttributeError: 'list' object has no attribute 'get'
```

Moving the first source into its own folder works again:

```yaml

  - git_url: https://github.com/jlumpe/lyve-SET.git
    git_rev: v1.1.4-head
    folder: lyve-set

  - git_url: https://github.com/lskatz/Schedule--SGELK.git
    folder: build/Schedule

  - git_url: https://github.com/lskatz/cg-pipeline
    folder: libs/cg-pipeline
```

All these examples happen when the only other contents of `meta.yaml` are `package` with `name` and `version` keys.

Running on Ubuntu 16.04. Info about my installation:

```
$ python --version
Python 3.5.2 :: Anaconda custom (64-bit)
$ conda --version
conda 4.3.22
$ conda-build --version
conda-build 3.0.8
```

**Comment by mandeep** [comment]
Could you show me the full contents of the meta.yaml?

**Comment by jlumpe** [comment]
Here's a minimal example I came up with, using 3 random popular Python git repos:

```yaml
package:
  name: my-package
  version: "1.0"

source:
  - git_url: https://github.com/pallets/flask.git

  - git_url: https://github.com/requests/requests.git
    folder: folder1

  - git_url: https://github.com/scrapy/scrapy.git
    folder: folder2
```

Just putting this in its own directory and running `conda-build .` generates the error.

Here is the full output:

```
Attempting to finalize metadata for my-package
INFO:conda_build.metadata:Attempting to finalize metadata for my-package
BUILD START: ['my-package-1.0-h904e138_0.tar.bz2']
Cloning into '/home/jared/anaconda3/conda-bld/my-package_1501259274309/work'...
done.
checkout: 'HEAD'
Your branch is up-to-date with 'origin/_conda_cache_origin_head'.
==> git log -n1 <==

commit 417bc336de7350f91378183e88cb35c1ef8c75c5
Author: David Lord <davidism@gmail.com>
Date:   Thu Jul 27 07:32:03 2017 -0700

    egg_info -R flag is gone

==> git describe --tags --dirty <==

0.12.2-414-g417bc33

==> git status <==

From https://github.com/requests/requests
 * [new ref]                    -> _conda_cache_origin_head
Cloning into '/home/jared/anaconda3/conda-bld/my-package_1501259274309/work/folder1'...
done.
On branch _conda_cache_origin_head
Your branch is up-to-date with 'origin/_conda_cache_origin_head'.
nothing to commit, working directory clean

checkout: 'HEAD'
Your branch is up-to-date with 'origin/_conda_cache_origin_head'.
==> git log -n1 <==

commit fdf426125b489c05b87c98a7141895aa3ade86ec
Merge: 12c7456 d6b57c6
Author: Nate Prewitt <nateprewitt@users.noreply.github.com>
Date:   Thu Jul 27 14:54:09 2017 -0600

    Merge pull request #4182 from alexwlchan/idna-version
    
    Add idna version info to requests.help

==> git describe --tags --dirty <==

v2.18.2-13-gfdf4261

==> git status <==

From https://github.com/scrapy/scrapy
 * [new ref]                    -> _conda_cache_origin_head
Cloning into '/home/jared/anaconda3/conda-bld/my-package_1501259274309/work/folder2'...
done.
On branch _conda_cache_origin_head
Your branch is up-to-date with 'origin/_conda_cache_origin_head'.
nothing to commit, working directory clean

checkout: 'HEAD'
Your branch is up-to-date with 'origin/_conda_cache_origin_head'.
==> git log -n1 <==

commit 9d34b2fe166f34400b15b21e95e37cb5a62aff6d
Merge: 5d9bac7 bf7ef3e
Author: Mikhail Korobov <kmike84@gmail.com>
Date:   Fri Jul 28 00:02:06 2017 +0500

    Merge pull request #2857 from redapple/scrapy-components-logs-startup
    
    Log versions information at startup

==> git describe --tags --dirty <==

1.4.0-63-g9d34b2f

==> git status <==

On branch _conda_cache_origin_head
Your branch is up-to-date with 'origin/_conda_cache_origin_head'.
nothing to commit, working directory clean

Traceback (most recent call last):
  File "/home/jared/anaconda3/bin/conda-build", line 6, in <module>
    sys.exit(conda_build.cli.main_build.main())
  File "/home/jared/anaconda3/lib/python3.5/site-packages/conda_build/cli/main_build.py", line 342, in main
    execute(sys.argv[1:])
  File "/home/jared/anaconda3/lib/python3.5/site-packages/conda_build/cli/main_build.py", line 333, in execute
    noverify=args.no_verify)
  File "/home/jared/anaconda3/lib/python3.5/site-packages/conda_build/api.py", line 184, in build
    need_source_download=need_source_download, config=config, variants=variants)
  File "/home/jared/anaconda3/lib/python3.5/site-packages/conda_build/build.py", line 1633, in build_tree
    notest=notest,
  File "/home/jared/anaconda3/lib/python3.5/site-packages/conda_build/build.py", line 999, in build
    m.parse_until_resolved(allow_no_other_outputs=True)
  File "/home/jared/anaconda3/lib/python3.5/site-packages/conda_build/metadata.py", line 824, in parse_until_resolved
    bypass_env_check=bypass_env_check)
  File "/home/jared/anaconda3/lib/python3.5/site-packages/conda_build/metadata.py", line 759, in parse_again
    bypass_env_check=bypass_env_check),
  File "/home/jared/anaconda3/lib/python3.5/site-packages/conda_build/metadata.py", line 1277, in _get_contents
    bypass_env_check=bypass_env_check))
  File "/home/jared/anaconda3/lib/python3.5/site-packages/conda_build/jinja_context.py", line 412, in context_processor
    ctx = get_environ(config=config, m=initial_metadata, for_env=False)
  File "/home/jared/anaconda3/lib/python3.5/site-packages/conda_build/environ.py", line 257, in get_dict
    d.update(meta_vars(m, config))
  File "/home/jared/anaconda3/lib/python3.5/site-packages/conda_build/environ.py", line 383, in meta_vars
    git_url = meta.get_value('source/git_url')
  File "/home/jared/anaconda3/lib/python3.5/site-packages/conda_build/metadata.py", line 891, in get_value
    value = self.get_section(section).get(key, default)
AttributeError: 'list' object has no attribute 'get'
```

**Comment by mandeep** [comment]
Thanks for the example. The reason why that one doesn't work is because the flask url does not specify a folder. All git_urls will need a folder specified as mentioned here: https://conda.io/docs/building/meta-yaml.html#source-section

**Comment by jlumpe** [comment]
Thank you for the help mandeep. Putting all git sources into their own folders was how I initially got around this problem, but unfortunately it fails again if I add a local source into the mix (even if it goes into its own folder as well):

```yaml
package:
  name: my-package
  version: "1.0"

source:
  - git_url: https://github.com/pallets/flask.git
    folder: folder0

  - git_url: https://github.com/requests/requests.git
    folder: folder1

  - git_url: https://github.com/scrapy/scrapy.git
    folder: folder2

  - path: src
    folder: folder3
```

```
Attempting to finalize metadata for my-package
INFO:conda_build.metadata:Attempting to finalize metadata for my-package
BUILD START: ['my-package-1.0-h16accef_0.tar.bz2']
Cloning into '/home/jared/anaconda3/conda-bld/my-package_1501278665160/work/folder0'...
done.
checkout: 'HEAD'
Your branch is up-to-date with 'origin/_conda_cache_origin_head'.
==> git log -n1 <==

commit 417bc336de7350f91378183e88cb35c1ef8c75c5
Author: David Lord <davidism@gmail.com>
Date:   Thu Jul 27 07:32:03 2017 -0700

    egg_info -R flag is gone

==> git describe --tags --dirty <==

0.12.2-414-g417bc33

==> git status <==

Cloning into '/home/jared/anaconda3/conda-bld/my-package_1501278665160/work/folder1'...
done.
On branch _conda_cache_origin_head
Your branch is up-to-date with 'origin/_conda_cache_origin_head'.
nothing to commit, working directory clean

checkout: 'HEAD'
Your branch is up-to-date with 'origin/_conda_cache_origin_head'.
==> git log -n1 <==

commit fdf426125b489c05b87c98a7141895aa3ade86ec
Merge: 12c7456 d6b57c6
Author: Nate Prewitt <nateprewitt@users.noreply.github.com>
Date:   Thu Jul 27 14:54:09 2017 -0600

    Merge pull request #4182 from alexwlchan/idna-version
    
    Add idna version info to requests.help

==> git describe --tags --dirty <==

v2.18.2-13-gfdf4261

==> git status <==

Cloning into '/home/jared/anaconda3/conda-bld/my-package_1501278665160/work/folder2'...
done.
On branch _conda_cache_origin_head
Your branch is up-to-date with 'origin/_conda_cache_origin_head'.
nothing to commit, working directory clean

checkout: 'HEAD'
Your branch is up-to-date with 'origin/_conda_cache_origin_head'.
==> git log -n1 <==

commit 9d34b2fe166f34400b15b21e95e37cb5a62aff6d
Merge: 5d9bac7 bf7ef3e
Author: Mikhail Korobov <kmike84@gmail.com>
Date:   Fri Jul 28 00:02:06 2017 +0500

    Merge pull request #2857 from redapple/scrapy-components-logs-startup
    
    Log versions information at startup

==> git describe --tags --dirty <==

1.4.0-63-g9d34b2f

==> git status <==

On branch _conda_cache_origin_head
Your branch is up-to-date with 'origin/_conda_cache_origin_head'.
nothing to commit, working directory clean

Traceback (most recent call last):
  File "/home/jared/anaconda3/bin/conda-build", line 6, in <module>
    sys.exit(conda_build.cli.main_build.main())
  File "/home/jared/anaconda3/lib/python3.5/site-packages/conda_build/cli/main_build.py", line 342, in main
    execute(sys.argv[1:])
  File "/home/jared/anaconda3/lib/python3.5/site-packages/conda_build/cli/main_build.py", line 333, in execute
    noverify=args.no_verify)
  File "/home/jared/anaconda3/lib/python3.5/site-packages/conda_build/api.py", line 184, in build
    need_source_download=need_source_download, config=config, variants=variants)
  File "/home/jared/anaconda3/lib/python3.5/site-packages/conda_build/build.py", line 1633, in build_tree
    notest=notest,
  File "/home/jared/anaconda3/lib/python3.5/site-packages/conda_build/build.py", line 996, in build
    try_download(m, no_download_source=False)
  File "/home/jared/anaconda3/lib/python3.5/site-packages/conda_build/render.py", line 355, in try_download
    source.provide(metadata)
  File "/home/jared/anaconda3/lib/python3.5/site-packages/conda_build/source.py", line 608, in provide
    path = normpath(abspath(join(metadata.path, metadata.get_value('source/path'))))
  File "/home/jared/anaconda3/lib/python3.5/site-packages/conda_build/metadata.py", line 891, in get_value
    value = self.get_section(section).get(key, default)
AttributeError: 'list' object has no attribute 'get'
```

The directory "src" exists in the same directory as the meta.yaml file and contains a single text file with the contents "test".

**Comment by mandeep** [comment]
Thanks for reporting this. Seems to be a bug in how path is parsed when source is given other arguments as well.

**Comment by msarahan** [comment]
Should be fixed by #2247 - thanks @stuarteberg.  I'll try to get a release out tomorrow.
